File: typescript/src/load.ts
import { getContainer, getAzureOpenAIClient, generateEmbeddings } from "./utils";
import * as fs from "fs";

const file = "../movies.json";

async function load() {
    try {
        const container = getContainer();
        const azureOpenAIClient = getAzureOpenAIClient();

        // Read movies.json file
        const movies = await readMoviesFromFile(file);

        // Load each movie into Cosmos DB container
        for (const movie of movies) {
            const movieWithEmbedding = await createMovieWithEmbedding(movie, azureOpenAIClient);
            await container.items.upsert(movieWithEmbedding);
            console.log(`Inserted movie: ${movie.title}`);
        }
    } catch (err) {
        console.error("Error loading movies:", err);
        process.exit(1);
    }
}

async function readMoviesFromFile(filePath: string): Promise<Movie[]> {
    try {
        console.log(`Reading movies from file: ${filePath}`);
        const data = await fs.promises.readFile(filePath, "utf-8");
        return JSON.parse(data) as Movie[];
    } catch (err) {
        throw new Error(`Error reading or parsing file: ${err}`);
    }
}

async function createMovieWithEmbedding(movie: Movie, client: any): Promise<any> {
    const embedding = await generateEmbeddings(movie.description, client);
    console.log(`Created embedding for movie: ${movie.title}`);
    return {
        id: movie.title,
        description: movie.description,
        embeddings: embedding,
    };
}

type Movie = {
    title: string,
    description: string
}

load();

File: python/load.py
from azure.cosmos import ContainerProxy
import utils
import json
import os

file_path = os.getenv("MOVIES_FILE_PATH", "../movies.json")

def main(container: ContainerProxy) -> None:
    with open(file_path, 'r') as file:
        movies = json.load(file)

    for movie in movies:
        try:
            title = movie["title"]
            description = movie["description"]
            embedding = utils.generate_embeddings(description)
            
            print("Generated description embedding for movie:", title)

            container.upsert_item({
                "id": title,
                #"title": title,
                "description": description,
                "embeddings": embedding
            })

            print("Added data to Cosmos DB for movie:", title)

        except Exception as e:
            print(f"Error processing movie {movie.get('title', 'unknown')}: {e}")

    print("Data loading complete")

if __name__ == "__main__":
    cosmos_container: ContainerProxy = utils.get_container()
    main(cosmos_container)